File: Backend/models/receta.js
const db = require("../config/db");

// ----------------------------------------
//     CLASE QUE REPRESENTA UNA RECETA
// ----------------------------------------
class RecetaEntity {
  constructor(obj) {
    this.id = obj.Id_receta;
    this.titulo = obj.titulo;
    this.duracion = obj.tiempo_preparacion;
    this.pais = obj.origen;
    this.alergenos = obj.alergenos;
    this.estacion = obj.estacion;
    this.publica = obj.publica; // convertimos a booleano
    this.usuarioId = obj.Id_usuario;
    this.ingredientes = obj.ingredientes || [];
    this.pasos = obj.pasos || [];
    this.imagen = obj.imagen || null; // base64 o null
    this.creadorNombre = obj.creadorNombre;
  }
}

// ----------------------------------------
//     MÉTODOS DE ACCESO A DATOS (MODELO)
// ----------------------------------------
const RecetaModel = {
  /**
   * Obtiene todas las recetas de un usuario (públicas + privadas)
   * Usado principalmente en el perfil del usuario
   */
  obtenerPorUsuario: async (userId) => {
    const [rows] = await db.query(
      `SELECT 
                r.Id_receta,
                r.titulo,
                i.imagen
            FROM receta r
            LEFT JOIN receta_imagen i ON i.Id_receta = r.Id_receta
            WHERE r.Id_usuario = ?`,
      [userId]
    );

    // Transformamos las filas en formato más amigable para el frontend
    const recetas = rows.map(row => ({
            id: row.Id_receta,
            titulo: row.titulo,
            imagenBase64: row.imagen ? `data:image/jpeg;base64,${row.imagen.toString('base64')}` : null
        }));

        return recetas;
    },

  /**
   * Obtiene recetas visibles (públicas)
   * Si se pasa userId → también incluye las recetas privadas de ese usuario
   */
  obtenerVisibles: async (userId = null) => {
    let query = "SELECT * FROM receta WHERE publica = 1";
    let params = [];

    if (userId) {
      query = "SELECT * FROM receta WHERE publica = 1 OR Id_usuario = ?";
      params = [userId];
    }

    const [rows] = await db.query(query, params);
    return rows.map(r => new RecetaEntity(r));
  },

  /**
   * Obtiene un rango de recetas (principalmente para la home / carrusel)
   * Devuelve solo lo mínimo necesario: id, título e imagen
   */
  getByRange: async (minId, maxId) => {
    const [rows] = await db.query(
      `SELECT 
                r.Id_receta,
                r.titulo,
                i.imagen
            FROM receta r
            LEFT JOIN receta_imagen i ON i.Id_receta = r.Id_receta
            WHERE r.Id_receta BETWEEN ? AND ?
            ORDER BY r.Id_receta
            LIMIT 6`,
      [minId, maxId]
    );

    return rows.map(row => ({
      Id_receta: row.Id_receta,
      titulo: row.titulo,
      imagenBase64: row.imagen
        ? `data:image/jpeg;base64,${row.imagen.toString("base64")}`
        : null,
    }));
  },

  /**
   * Obtiene UNA receta completa por su ID (con ingredientes, pasos e imagen)
   * Es la consulta más completa y usada en la vista de detalle
   */
  obtenerPorId: async (id) => {
    console.log("🔎 Consultando receta en DB, Id:", id);

    const [rows] = await db.query('SELECT r.*, u.nombre AS creadorNombre FROM receta r JOIN usuario u ON u.Id_usuario = r.Id_usuario WHERE r.Id_receta = ?', [
      id,
    ]);
    if (rows.length === 0) {
      console.log("⚠️ No se encontró la receta en DB");
      return null;
    }

    const receta = rows[0];

    // ── Ingredientes ───────────────────────────────
    const [ingredientes] = await db.query(
      "SELECT nombre, cantidad FROM Ingrediente WHERE Id_receta = ?",
      [id]
    );

    // ── Pasos ──────────────────────────────────────
    const [pasos] = await db.query(
      "SELECT descripcion FROM Paso WHERE Id_receta = ?",
      [id]
    );

    // ── Imagen (solo la primera por ahora) ─────────
    const [imagenes] = await db.query(
      "SELECT imagen FROM receta_imagen WHERE Id_receta = ?",
      [id]
    );

    // Armamos el objeto final
    receta.ingredientes = ingredientes;
    receta.pasos = pasos; // solo las descripciones
    receta.imagen = imagenes.length
      ? `data:image/jpeg;base64,${imagenes[0].imagen.toString("base64")}`
      : null;

    return new RecetaEntity(receta);
  },

  /**
   * Crea una receta nueva + sus relaciones (imagen, pasos, ingredientes)
   * @returns {number} ID de la receta recién creada
   */
  crear: async (data, userId) => {
    const [result] = await db.query(
      `INSERT INTO Receta 
            (titulo, tiempo_preparacion, origen, alergenos, estacion, publica, Id_usuario) 
            VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [
        data.titulo,
        data.duracion,
        data.pais,
        data.alergenos,
        data.estacion,
        0,
        userId,
      ]
    );

    const recetaId = result.insertId;

    // Imagen (si existe)
    if (data.imagen) {
      const base64Data = data.imagen.replace(/^data:image\/\w+;base64,/, "");
      const buffer = Buffer.from(base64Data, "base64");
      await db.query(
        "INSERT INTO receta_imagen (imagen, Id_receta) VALUES (?, ?)",
        [buffer, recetaId]
      );
    }

    // Pasos
    if (data.pasos?.length) {
      const pasosPromises = data.pasos.map(p =>
        db.query("INSERT INTO Paso (descripcion, Id_receta) VALUES (?, ?)", [
          p.descripcion,
          recetaId,
        ])
      );
      await Promise.all(pasosPromises);
    }

    // Ingredientes
    if (data.ingredientes?.length) {
      const ingredientesPromises = data.ingredientes.map(i =>
        db.query(
          "INSERT INTO Ingrediente (nombre, cantidad, Id_receta) VALUES (?, ?, ?)",
          [i.nombre, i.cantidad, recetaId]
        )
      );
      await Promise.all(ingredientesPromises);
    }

    return recetaId;
  },

  /**
   * Actualización simple (solo campos principales de la tabla Receta)
   * Nota: no actualiza ingredientes ni pasos en esta implementación
   */
  actualizar: async (id, data) => {
    await db.query("UPDATE Receta SET ? WHERE Id_receta = ?", [data, id]);
  },

  /**
   * Eliminación física de la receta
   * (En producción se recomienda más bien un borrado lógico con campo activo/eliminado)
   */
  eliminar: async (id) => {
    await db.query("DELETE FROM Receta WHERE Id_receta = ?", [id]);
  },

  /**
   * Verifica si un usuario es el propietario de una receta
   * @returns {boolean|null} true = es propietario, false = no lo es, null = receta no existe
   */
  verificarPropietario: async (id, userId) => {
    const [rows] = await db.query(
      "SELECT Id_usuario FROM Receta WHERE Id_receta = ?",
      [id]
    );
    if (rows.length === 0) return null;
    return rows[0].Id_usuario === userId;
  },

  recetasFiltradas: async (filtros = {}) => {
    let query = `
      SELECT DISTINCT
        r.*,
        ri.imagen
      FROM receta r
      LEFT JOIN Ingrediente i ON i.Id_receta = r.Id_receta
      LEFT JOIN receta_imagen ri ON ri.Id_receta = r.Id_receta
    `;
    let where = [];
    let params = [];

    // 🔍 Texto libre → título + ingredientes
    if (filtros.texto) {
      const palabras = filtros.texto.split(" ").filter(t => t.trim() !== "");
      if (palabras.length) {
        const likeClauses = palabras
          .map(() => "(r.titulo LIKE ? OR i.nombre LIKE ?)")
          .join(" AND ");
        where.push(likeClauses);
        palabras.forEach(p => {
          params.push(`%${p}%`, `%${p}%`); // una vez para titulo, otra para ingrediente
        });
      }
    }

    // 🌍 País / origen
    if (filtros.pais) {
      where.push("r.origen = ?");
      params.push(filtros.pais);
    }

    // ⏱️ Duración máxima
    if (filtros.duracion) {
      where.push("r.tiempo_preparacion <= ?");
      params.push(filtros.duracion);
    }

    // 🌦️ Estación
    if (filtros.estacion) {
      where.push("r.estacion = ?");
      params.push(filtros.estacion);
    }

    // ⚠️ Alérgenos (si es string o array)
    if (filtros.alergenos?.length) {
      filtros.alergenos.forEach(a => {
        where.push("LOWER(r.alergenos) NOT LIKE ?");
        params.push(`%${a.toLowerCase()}%`);
      });
    }

    // 👁️ Visibilidad
    if (filtros.userId) {
      // Públicas + privadas del propio usuario
      where.push("(r.publica = 1 OR r.Id_usuario = ?)");
      params.push(filtros.userId);
    } else {
      // Usuario no autenticado → solo públicas
      where.push("r.publica = 1");
    }

    // 🧩 Unimos condiciones
    if (where.length) {
      query += " WHERE " + where.join(" AND ");
    }

    // 🔽 Orden opcional
    query += " ORDER BY r.Id_receta DESC";

    //Paginación
    const limit = parseInt(filtros.limit) || 20;
    const offset = parseInt(filtros.offset) || 0;
    query += " LIMIT ? OFFSET ?";
    params.push(limit, offset);

    const [rows] = await db.query(query, params);
    return rows.map(r => {
      return new RecetaEntity({
        ...r,
        imagen: r.imagen
          ? `data:image/webp;base64,${r.imagen.toString('base64')}`
          : null,
      });
    });
  }
};

module.exports = { RecetaEntity, RecetaModel };

File: remy_recipes/lib/data/models/receta.dart
import '../constants/app_strings.dart';

// ==========================================================================
// 1. CLASES MODELO AUXILIARES
// ==========================================================================

class Ingrediente {
  String nombre;
  String cantidad;
  Ingrediente({required this.nombre, required this.cantidad});
  Map<String, dynamic> toJson() => {'nombre': nombre, 'cantidad': cantidad};
}

class Paso {
  String descripcion;
  Paso({required this.descripcion});
  Map<String, dynamic> toJson() => {'descripcion': descripcion};
}

class Receta {
  final String? id;
  final String titulo;
  final List<Ingrediente>? ingredientes;
  final List<Paso>? pasos;
  final int? duracion;
  final String? pais;
  final String? alergenos;
  final String? estacion;
  final int? idUsuario;
  final String? imagenBase64;
  final String? creadorNombre;

  Receta({
    this.id,
    required this.titulo,
    this.ingredientes,
    this.pasos,
    this.duracion,
    this.pais,
    this.alergenos,
    this.estacion,
    this.idUsuario,
    this.imagenBase64,
    this.creadorNombre,
  });

  // Constructor factory para crear un objeto Receta desde un JSON (GET)

  factory Receta.fromJson(Map<String, dynamic> json) {
    return Receta(
      id: json['Id_receta']?.toString() ?? json['id']?.toString(),
      titulo: json['titulo']?.toString() ?? AppStrings.sinTitulo,
      ingredientes: (json['ingredientes'] as List<dynamic>?)
          ?.map(
            (i) => Ingrediente(nombre: i['nombre'], cantidad: i['cantidad']),
          )
          .toList(),
      pasos: (json['pasos'] as List<dynamic>?)
          ?.map((p) => Paso(descripcion: p['descripcion']))
          .toList(),
      duracion: json['duracion'] != null
          ? int.parse(json['duracion'].toString())
          : null,
      pais: json['pais']?.toString(),
      alergenos: json['alergenos']?.toString(),
      estacion: json['estacion']?.toString(),
      idUsuario: json['usuarioId'] as int?,
      imagenBase64: json['imagen']?.toString(),
      creadorNombre: json['creadorNombre']?.toString(),
    );
  }

  // PARA HOME (GRID)
  factory Receta.fromHomeJson(Map<String, dynamic> json) {
    print('🟢 fromHomeJson recibido: $json');
    return Receta(
      id: json['id'] != null
          ? json['id'].toString()
          : (json['Id_receta']?.toString()),
      titulo: json['titulo']?.toString() ?? 'Sin título',
      imagenBase64:
          json['imagenBase64']?.toString() ?? json['imagen']?.toString(),
    );
  }

  // Método para convertir el objeto Receta a un mapa JSON (POST/PUT)

  Map<String, dynamic> toJson() {
    return {
      'titulo': titulo,
      //  lista de JSONs que el JS espera
      if (ingredientes != null)
        'ingredientes': ingredientes!.map((i) => i.toJson()).toList(),
      if (pasos != null) 'pasos': pasos!.map((p) => p.toJson()).toList(),
      if (duracion != null) 'duracion': duracion,
      if (pais != null) 'pais': pais,
      if (alergenos != null) 'alergenos': alergenos,
      if (estacion != null) 'estacion': estacion,
      if (idUsuario != null) 'Id_usuario': idUsuario,
      if (imagenBase64 != null) 'imagen': imagenBase64,
    };
  }
}
